### PR TITLE
Add support for post-scripts like postinstall, poststart, postbuild etc.

### PR DESCRIPTION
**Pull request by b4s36t4**
#86

### CODE REVIEW

**Comment by b4s36t4** [comment]
@danielhuang can you review this if possible?

File: src/main.rs [hunk 1]
```diff
@@ -129,6 +129,38 @@ pub async fn verify_installation(package: &Package, plan: &Plan) -> Result<bool>
     Ok(installed.satisfies(package))
 }
 
+async fn post_script(name: &str) -> Result<()> {
+    let package = read_package().await?;
+
+    join_paths()?;
+
+    let script_key: &str = &("post".to_owned() + &name.to_owned());
```
Comment (danielhuang) [inline src/main.rs:137]: This can be replaced with `format!`

File: src/main.rs [hunk 2]
```diff
@@ -129,6 +129,38 @@ pub async fn verify_installation(package: &Package, plan: &Plan) -> Result<bool>
...
+
+    if !package.scripts.contains_key(script_key) {
+        return Ok(());
+    }
+
+    PROGRESS_BAR.suspend(|| println!("Running {}", script_key));
```
Comment (danielhuang) [inline src/main.rs:143]: `PROGRESS_BAR` has a `println` method. `.suspend` is not needed here.

File: src/main.rs [hunk 3]
```diff
@@ -129,6 +129,38 @@ pub async fn verify_installation(package: &Package, plan: &Plan) -> Result<bool>
...
+
+    let mut child = Command::new(&shell().await?)
+        .arg("-c")
+        .arg(script)
+        .spawn()?;
+
+    let exit_code = child.wait().await?.code();
+
+    if let Some(exit_code) = exit_code {
+        exit(exit_code);
```
Comment (danielhuang) [inline src/main.rs:159]: This will exit the entire process.

**Review by danielhuang — CHANGES_REQUESTED** [review]
This is a good start! However, it only supports the user-specified `postinstall` script in the project itself, not the necessary scripts used by some packages used as dependencies. 

The installation process currently works like this:
1. Make sure that the intended layout of `node_modules` is determined. If there are packages needed and the metadata is not yet known, fetch the metadata from the registry.
2. For each intended package, begin downloading the tarball that contains the files inside that package. The first 2 steps will happen concurrently, by starting the download process immediately after the metadata is known.
3. Once step 1 is complete, begin constructing `node_modules`. It would go through each package, wait for the download process (if needed), and hardlink the files to the target location.

In order to fully support `postinstall` scripts, certain dependencies needs to be completed before executing the intended scripts. For example, `electron` uses `pump` (a transitive dependency) as part of its `postinstall` in order to download a compiled artifact. That would need `electron`'s `postinstall` to wait on all of its dependencies.

**Comment by danielhuang** [comment]
Install scripts is now supported. See #86.

**Pull request closed without merging**